From a203bba4dc69375947d4bb5ecefd7932b95ff41f Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@users.sf.net>
Date: Thu, 26 Jan 2012 16:56:42 +0100
Subject: [PATCH 2/6] Allow wrapping of vtkRenderWindow::GetColorBufferSizes

See http://www.vtk.org/Bug/view.php?id=7857

When color depth is lower than 24, vtkVisibleCellSelector::Select()
silently fails.  We want to detect this case to provide
a software alternative, or at least tell user why picking
does not work.
---
 Rendering/vtkMesaRenderWindow.h     |    1 +
 Rendering/vtkOpenGLRenderWindow.cxx |   17 +++++++++++++++++
 Rendering/vtkOpenGLRenderWindow.h   |    1 +
 Rendering/vtkRenderWindow.h         |    2 ++
 4 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/Rendering/vtkMesaRenderWindow.h b/Rendering/vtkMesaRenderWindow.h
index 17bd437..ff5223b 100644
--- a/Rendering/vtkMesaRenderWindow.h
+++ b/Rendering/vtkMesaRenderWindow.h
@@ -108,6 +108,7 @@ public:
   // Get the size of the color buffer.
   // Returns 0 if not able to determine otherwise sets R G B and A into buffer.
   int GetColorBufferSizes(int *rgba);
+  int GetColorBufferSizes(vtkIntArray *rgba);
 
   // Description:
   // Initialize OpenGL for this window.
diff --git a/Rendering/vtkOpenGLRenderWindow.cxx b/Rendering/vtkOpenGLRenderWindow.cxx
index ea50038..27a099e 100644
--- a/Rendering/vtkOpenGLRenderWindow.cxx
+++ b/Rendering/vtkOpenGLRenderWindow.cxx
@@ -17,6 +17,7 @@
 #include "vtkFloatArray.h"
 #include "vtkgl.h"
 #include "vtkIdList.h"
+#include "vtkIntArray.h"
 #include "vtkObjectFactory.h"
 #include "vtkOpenGLActor.h"
 #include "vtkOpenGLCamera.h"
@@ -378,6 +379,22 @@ int vtkOpenGLRenderWindow::GetColorBufferSizes(int *rgba)
     }
 }
 
+int vtkOpenGLRenderWindow::GetColorBufferSizes(vtkIntArray *rgba)
+{
+  if (rgba==NULL)
+    {
+    return 0;
+    }
+
+  if ( rgba->GetMaxId()+1 != 4)
+    {
+    vtkDebugMacro("Resizing array.");
+    rgba->SetNumberOfComponents(1);
+    rgba->SetNumberOfValues(4);
+    }
+  return this->GetColorBufferSizes(rgba->GetPointer(0));
+}
+
 unsigned char* vtkOpenGLRenderWindow::GetPixelData(int x1, int y1, 
                                                    int x2, int y2,
                                                    int front)
diff --git a/Rendering/vtkOpenGLRenderWindow.h b/Rendering/vtkOpenGLRenderWindow.h
index c36602d..9e46001 100644
--- a/Rendering/vtkOpenGLRenderWindow.h
+++ b/Rendering/vtkOpenGLRenderWindow.h
@@ -102,6 +102,7 @@ public:
   // Get the size of the color buffer.
   // Returns 0 if not able to determine otherwise sets R G B and A into buffer.
   int GetColorBufferSizes(int *rgba);
+  int GetColorBufferSizes(vtkIntArray *rgba);
 
   // Description:
   // Initialize OpenGL for this window.
diff --git a/Rendering/vtkRenderWindow.h b/Rendering/vtkRenderWindow.h
index c61b185..5a6c248 100644
--- a/Rendering/vtkRenderWindow.h
+++ b/Rendering/vtkRenderWindow.h
@@ -41,6 +41,7 @@
 #include "vtkWindow.h"
 
 class vtkFloatArray;
+class vtkIntArray;
 class vtkPainterDeviceAdapter;
 class vtkRenderWindowInteractor;
 class vtkRenderer;
@@ -512,6 +513,7 @@ public:
   // Get the size of the color buffer.
   // Returns 0 if not able to determine otherwise sets R G B and A into buffer.
   virtual int GetColorBufferSizes(int *rgba) = 0;
+  virtual int GetColorBufferSizes(vtkIntArray *rgba) = 0;
 
   // Description:
   // Get the vtkPainterDeviceAdapter which can be used to paint on
-- 
1.7.8.3

